File: examples/go/create_vm.go
package examplesgo

import (
	"context"
	"errors"
)

var ErrDeadline = errors.New("no completion response before deadline")
var ErrUnresolved = errors.New("logical operation unresolved")

type VM struct{ Size string }
type Client interface {
	CreateVM(context.Context, string, VM) error
}

// The blank identity models a receiver with no documented repeat protection.
// faultscope:begin fs-c01.retry-independent-attempt
func RetryIndependent(ctx context.Context, client Client, spec VM) error {
	err := client.CreateVM(ctx, "", spec)
	if errors.Is(err, ErrDeadline) {
		return client.CreateVM(ctx, "", spec)
	}
	return err
}

// faultscope:end fs-c01.retry-independent-attempt

// faultscope:begin fs-c01.keep-unresolved
func KeepUnresolved(err error) error {
	if errors.Is(err, ErrDeadline) {
		return ErrUnresolved // No response does not establish no VM.
	}
	return err
}

// faultscope:end fs-c01.keep-unresolved

// The stronger synthetic receiver contract binds compatible repeats to P.
// faultscope:begin fs-c01.retry-same-logical-operation
func RetrySameOperation(ctx context.Context, client Client, spec VM, p string) error {
	err := client.CreateVM(ctx, p, spec)
	if errors.Is(err, ErrDeadline) {
		return client.CreateVM(ctx, p, spec)
	}
	return err
}

// faultscope:end fs-c01.retry-same-logical-operation

// faultscope:begin fs-c01.retry-with-new-logical-operation
func RetryWithNewOperation(ctx context.Context, client Client, spec VM, p, q string) error {
	err := client.CreateVM(ctx, p, spec)
	if errors.Is(err, ErrDeadline) {
		return client.CreateVM(ctx, q, spec) // Q is another logical operation.
	}
	return err
}

// faultscope:end fs-c01.retry-with-new-logical-operation

File: examples/c/create_vm.c
#include <stddef.h>

typedef enum { VM_OK, VM_NO_COMPLETION_RESPONSE, VM_OTHER_ERROR, VM_UNRESOLVED } VmOutcome;
typedef struct { const char *size; } VmSpec;
typedef VmOutcome (*CreateVm)(void *context, const char *operation_id, VmSpec spec);
typedef struct { void *context; CreateVm create_vm; } VmClient;

// faultscope:begin fs-c01.retry-independent-attempt
VmOutcome retry_independent(VmClient client, VmSpec spec) {
    VmOutcome result = client.create_vm(client.context, NULL, spec);
    if (result == VM_NO_COMPLETION_RESPONSE) {
        return client.create_vm(client.context, NULL, spec); // No repeat protection is documented.
    }
    return result;
}
// faultscope:end fs-c01.retry-independent-attempt

// faultscope:begin fs-c01.keep-unresolved
VmOutcome keep_unresolved(VmOutcome result) {
    return result == VM_NO_COMPLETION_RESPONSE ? VM_UNRESOLVED : result;
}
// faultscope:end fs-c01.keep-unresolved

// faultscope:begin fs-c01.retry-same-logical-operation
VmOutcome retry_same_operation(VmClient client, VmSpec spec, const char *p) {
    VmOutcome result = client.create_vm(client.context, p, spec);
    if (result == VM_NO_COMPLETION_RESPONSE) {
        return client.create_vm(client.context, p, spec); // Same P under stronger contract.
    }
    return result;
}
// faultscope:end fs-c01.retry-same-logical-operation

// faultscope:begin fs-c01.retry-with-new-logical-operation
VmOutcome retry_with_new_operation(VmClient client, VmSpec spec, const char *p, const char *q) {
    VmOutcome result = client.create_vm(client.context, p, spec);
    if (result == VM_NO_COMPLETION_RESPONSE) {
        return client.create_vm(client.context, q, spec); // Q is another logical operation.
    }
    return result;
}
// faultscope:end fs-c01.retry-with-new-logical-operation
